src/cmd/ksh93/tests/special-dev-paths.sh

```diff
@@ -0,0 +1,157 @@
+#
+# Verify that the special /dev paths such as /dev/stdin, /dev/fd/0, and /dev/tcp/host/port work when
+# used in a test and actually opened. This is primarily meant to exercise the `sh_open()` functions
+# handling of special paths.
+#
+readonly host=localhost
+readonly port=65123
+
+# Some tests will want to verify whether or not a fd was allocated. We could use the `lsof` command
+# but don't want to add that as a dependency. Instead just do the verification on platforms which
+# support /proc/self/fd.
+[[ -d /proc/self/fd ]] && readonly fd_check=yes || readonly fd_check=no
+[[ $fd_check == yes ]] && readonly fd_count=$( ls -1 /proc/$$/fd | wc -l )
+
+# Check if the `nc` (netcat) command is available.
+whence -q nc && readonly nc_available=yes || readonly nc_available=no
+
+# Check if the system supports the /dev/{stdin,stdout,stderr} paths.
+[[ -e /dev/stdin ]] && readonly dev_stdin=yes || readonly dev_stdin=no
+
+# ==========
+# Verify that testing variations of the special dev paths work as expected.
+# Note that these should not result in a fd being opened.
+[[ -e /dev/stdin ]] || log_error '-e /dev/stdin failed'
+[[ -e /dev/stdout ]] || log_error '-e /dev/stdout failed'
+[[ -e /dev/stderr ]] || log_error '-e /dev/stderr failed'
+[[ -e /dev/stdinX ]] && log_error '-e /dev/stdinX succeeded but should have failed'
+[[ -e /dev/stdoutX ]] && log_error '-e /dev/stdoutX succeeded but should have failed'
+[[ -e /dev/stderrX ]] && log_error '-e /dev/stderrX succeeded but should have failed'
+
+[[ -e /dev/fd/0 ]] || log_error '-e /dev/fd/0 failed'
+# TODO: This test currently fails. It thinks the bogus path is valid.
+# [[ -e /dev/fd/xxx ]] && log_error '-e /dev/fd/xxx succeeded but should have failed'
+
+[[ -e /dev/tcp/localhost/666 ]] || log_error '-e /dev/tcp/localhost/666 failed'
```

Comment:
This may be expected behavior, but why should this test succeed ?

Reply:
The original test in *brackets.sh* simply verified that `/dev/tcp/` is valid. Which could be valid or invalid depending on how you look at it. You can see from the other tests I've included but commented out with a "TODO" that I consider it undesirable because you can't actually do `echo hello >/dev/tcp` or `echo bye >/dev/tcp/localhost`. On the other hand if you treat this as if the */dev/tcp* directory actually existed and hostnames formed another layer of directories then `[[ -e /dev/tcp ]]` arguably should be true. Too, we don't want to go to the expense of validating the hostname is valid and we can validate the port number only to the extent it is in the range 0 to 65535. But we should be verifying that `/dev/tcp/localhost/1234/abc` is invalid.

Reply:
> This may be expected behavior, but why should this test succeed ?

I see I did not directly answer this question with my previous reply. That path is theoretically valid and thus the `[[ -e ... ]]` test should be true. The test cannot verify the path is actually valid since that can only be done by establishing a network connection. Something we don't want to do because the side-effects are undesirable. This is behavior that is insufficiently documented. What does it even mean to test if a */dev/tcp/...* path exists? Should it validate the path is valid if actually used to open a connection? Or should it just validate the path has the proper form?

Reply:
Ok, it would be good to mention these assumptions in comments with this test case.

Reply:
Will do.